taskcluster/translations_taskgraph/transforms/from_datasets.py

```diff
@@ -0,0 +1,155 @@
+            #   excluded. If specific pairs are listed, only those pairs are excluded.
+            Optional("exclude-datasets"): {
+                # provider name
+                str: {
+                    # dataset name
+                    str: [
+                        # src/trg locale pairs
+                        {
+                            "src": str,
+                            "trg": str,
+                        },
+                    ],
+                },
+            },
+            # Locales that will be excluded from all providers & datasets
+            Optional("exclude-locales"): [
+                {
+                    "src": str,
+                    "trg": str,
+                },
+            ],
+        },
+    },
+    extra=ALLOW_EXTRA,
+)
+
+transforms = TransformSequence()
+transforms.add_validate(SCHEMA)
+
+
+def substitute(item, **subs):
+    if isinstance(item, list):
+        for i in range(len(item)):
+            item[i] = substitute(item[i], **subs)
+    elif isinstance(item, dict):
+        new_dict = {}
+        for k, v in item.items():
+            k = k.format(**subs)
+            new_dict[k] = substitute(v, **subs)
+        item = new_dict
+    elif isinstance(item, str):
+        item = item.format(**subs)
+    else:
+        item = item
+
+    return item
+
+
+def shorten_dataset_name(dataset):
+    """Shortens various dataset names. Mainly used to make sure we can have
+    useful Treeherder symbols."""
+    # TODO: should the replacements live in ci/config.yml?
+    return (dataset
+        .replace("new-crawl", "nc")
+        .replace("news.2020", "n2020")
+        .replace("Neulab-tedtalks_train-1", "Ntt1")
+    )
+
+@transforms.add
+def jobs_from_datasets(config, jobs):
+    for job in jobs:
+        dataset_config = job.pop("dataset-config", {})
+        include_datasets = dataset_config.get("include-datasets", {})
+        exclude_datasets = dataset_config.get("exclude-datasets", {})
+        exclude_locales = dataset_config.get("exclude-locales", [])
+        substitution_fields = dataset_config.get("substitution-fields", [])
+
+        providers = include_datasets.keys()
+        if not providers:
+            providers = config.graph_config["datasets"].keys()
+
+        for provider in providers:
+            datasets = include_datasets.get(provider, {})
+            if not datasets:
+                datasets = config.graph_config["datasets"][provider]
+
+            for dataset, locale_pairs in datasets.items():
+                if dataset in exclude_datasets.get(provider, {}):
+                    continue
+
+                for pair in locale_pairs:
+                    if pair in exclude_datasets.get(provider, {}).get(dataset, []):
+                        continue
+
+                    if pair in exclude_locales:
+                        continue
+
+                    subjob = copy.deepcopy(job)
```

Comment:
Oh wow, this could fan out quite a lot! I glossed over that in the previous implementation. I see why it gets expensive.